feat: improve MCP config UI with comprehensive add/edit/delete functionality

xingyaoww opened the pull request:
- [x] This change is worth documenting at https://docs.all-hands.dev/
- [x] Include this change in the Release Notes. If checked, you **must** provide an **end-user friendly** description for your change below

**End-user friendly description of the problem this fixes or functionality this introduces.**

This PR significantly improves the MCP (Model Context Protocol) configuration experience in the OpenHands frontend. Previously, users had to manually edit JSON configuration files to manage MCP servers, which was error-prone and difficult for non-technical users. Now, users can easily add, edit, and delete MCP servers through an intuitive form-based interface with comprehensive validation and helpful error messages.

The new UI supports all three types of MCP servers (SSE, STDIO, SHTTP) and is available in both SaaS and OSS modes. Users can now configure MCP servers with confidence, knowing that the interface will guide them through the process and validate their inputs in real-time.

---
**Summarize what the PR does, explaining any non-trivial design decisions.**

This PR completely overhauls the MCP settings interface by:

1. **Replacing JSON editor with form-based UI**: Moved from a raw JSON editor to an intuitive form interface that follows the same patterns as the secrets settings page.

2. **Comprehensive server type support**: Added full support for all three MCP server types:
   - SSE (Server-Sent Events): URL and optional API key
   - STDIO (Standard I/O): Name, command, arguments, and environment variables
   - SHTTP (Secure HTTP): URL and optional API key

3. **Advanced form validation**: Implemented real-time validation including:
   - URL validation with protocol checking for SSE/SHTTP servers
   - Name validation (alphanumeric + hyphens/underscores) for STDIO servers
   - Command validation (no spaces allowed) for STDIO servers
   - Environment variable parsing and validation

4. **New reusable components**:
   - `MCPServerList`: Table view displaying all configured servers with edit/delete actions
   - `MCPServerForm`: Comprehensive form for adding/editing servers with type-specific fields

5. **Enhanced UX features**:
   - Confirmation dialogs for destructive actions
   - Dirty state tracking for unsaved changes
   - Loading states and proper error handling
   - Helpful placeholder text and validation messages

6. **Internationalization**: Added 29 new translation keys with support for all existing languages (en, ja, zh-CN, zh-TW, ko-KR, no, it, pt, es, ar, fr, tr, de, uk).

7. **Cross-mode availability**: Made MCP settings available in both SaaS and OSS modes by updating the navigation configuration.

**Design decisions:**
- Followed the existing secrets settings pattern for consistency
- Used TypeScript for type safety and better developer experience
- Implemented proper state management with React hooks
- Added comprehensive form validation to prevent configuration errors
- Used existing UI components and styling for visual consistency

---
**Link of any specific issues this addresses:**

This addresses the user request to improve the MCP config UI in the frontend by building a comprehensive form-based UI for adding/editing/deleting MCP configurations and displaying MCP settings in both SaaS and OSS modes.

xingyaoww commented:
@OpenHands please fix the failing actions on PR https://github.com/All-Hands-AI/OpenHands/pull/10145 on branch feature/improve-mcp-config-ui

Abubakar-01 commented:
**Before PR (main branch Behaviour):**
CLI and UI both lacked comprehensive add, edit, and delete functionality for MCP configuration.
Users could not fully manage MCP configs through either the CLI or the UI.

<img width="512" height="386" alt="image" src="https://github.com/user-attachments/assets/bd072c98-a9a0-4f05-8482-96654879c01d" />
<img width="1280" height="461" alt="image" src="https://github.com/user-attachments/assets/3d833579-0549-4183-bce5-15423161d0ab" />

**After PR (PR branch Behaviour):**
CLI now supports comprehensive add, edit, and delete functionality for MCP configuration.
UI still does not provide full add, edit, and delete capabilities for MCP configs.

**✅CLI**
<img width="1280" height="529" alt="image" src="https://github.com/user-attachments/assets/14a011bc-fd82-4203-9a9b-fa4a74b4428d" />
<img width="512" height="243" alt="image" src="https://github.com/user-attachments/assets/9b8d4917-a484-4a1f-8e08-22a133df6165" />

**❌UI**
<img width="1280" height="197" alt="image" src="https://github.com/user-attachments/assets/13c4b6a0-99ad-4ad5-bbd2-46f1c0f99c21" />


**Conclusion:**
The PR successfully adds comprehensive MCP config management to the CLI, but the UI remains incomplete in this regard. Further work is needed to bring the UI in line with the CLI’s new capabilities.

xingyaoww commented:
@Abubakar-01 can you comment out what is missing in particular in the UI so we can ask OpenHands to fix it?

Abubakar-01 commented:
Ui has no option to fix (delete edit) MCP config.

xingyaoww reviewed (commented):
@Abubakar-01 Tweak the code again and fix the remaining issue -- can you qa it again?


<img width="2404" height="850" alt="Image" src="https://github.com/user-attachments/assets/1fc560b7-b33c-4310-8135-95f0b11c327e" />

Abubakar-01 commented:
@xingyaoww  Yes I will test it again!

Abubakar-01 commented:
@xingyaoww  Overflow issue is fixed now

Abubakar-01 commented:
Here is the video
[Screencast from 12-08-2025 23:11:43.webm](https://github.com/user-attachments/assets/0d5fd67f-418b-4f71-abf7-796e0bc3dbf6)

xingyaoww commented:
Thank you!

hieptl commented:
suggestion(non-blocking): Should we make the `Server Type` dropdown match the width of the input fields? Please refer to the image below for more details.

<img width="1440" height="743" alt="suggestion-1" src="https://github.com/user-attachments/assets/4602383c-047e-4b54-840b-5000e7e2ed82" />

suggestion(non-blocking): For the `Server Details` column, should we truncate the text and display ... at the end? When users hover over the `Name` or `Server Details` columns, a tooltip could appear to show the full content. Please refer to the image below for more details.

<img width="1440" height="743" alt="suggestion-2" src="https://github.com/user-attachments/assets/2fd7db02-ca0f-4ce2-9911-11e76e6109a1" />

question(non-blocking): Do we allow duplicate records in this table? Please refer to the image below for more details.

<img width="1440" height="743" alt="question-1" src="https://github.com/user-attachments/assets/0690a923-145b-4b82-a747-016076e6bbbd" />

suggestion(non-blocking): I think we should validate the environment variables field. When I enter free text in that field without following the required format, for example:

```
test1
test2
test3
```

The app should display an alert indicating that environment variables must follow the format:

```
KEY1=value1
KEY2=value2
```

However, I’m currently able to save without any warning. Please refer to the video below for more details.

https://github.com/user-attachments/assets/9575e243-e823-4d98-8211-e12f64d7382d

hieptl commented on frontend/src/routes/mcp-settings.tsx:
```diff
@@ -1,86 +1,191 @@
 import { useSettings } from "#/hooks/query/use-settings";
-import { useSaveSettings } from "#/hooks/mutation/use-save-settings";
-import { MCPConfig } from "#/types/settings";
-import { MCPConfigEditor } from "#/components/features/settings/mcp-settings/mcp-config-editor";
```
**suggestion(non-blocking):** I see that this component is not used anywhere in the codebase. Should we remove it?

xingyaoww replied:
done!

hieptl commented on frontend/src/components/features/settings/mcp-settings/mcp-json-editor.tsx:
```diff
@@ -14,6 +14,7 @@ interface MCPJsonEditorProps {
+  shttp_servers: [],
 };
 
 export function MCPJsonEditor({
```
**suggestion (non-blocking):** Since we are not using the `<MCPConfigEditor />` component, and the `<MCPJsonEditor />` component is only used inside it, should we remove this one as well?

xingyaoww replied:
done

hieptl commented on frontend/src/components/features/settings/mcp-settings/mcp-server-form.tsx:
```diff
@@ -0,0 +1,321 @@
+    { key: "shttp", label: t(I18nKey.SETTINGS$MCP_SERVER_TYPE_SHTTP) },
+  ];
+
+  const validateForm = (formData: FormData): string | null => {
```
**suggestion (non-blocking):** This function can be broken down into smaller functions as follows.

```js
const validateUrl = (url: string): string | null => {
  if (!url) {
    return t(I18nKey.SETTINGS$MCP_ERROR_URL_REQUIRED);
  }

  try {
    const urlObj = new URL(url);
    if (!["http:", "https:"].includes(urlObj.protocol)) {
      return t(I18nKey.SETTINGS$MCP_ERROR_URL_INVALID_PROTOCOL);
    }
  } catch {
    return t(I18nKey.SETTINGS$MCP_ERROR_URL_INVALID);
  }

  return null;
};

const validateName = (name: string): string | null => {
  if (!name) {
    return t(I18nKey.SETTINGS$MCP_ERROR_NAME_REQUIRED);
  }

  if (!/^[a-zA-Z0-9_-]+$/.test(name)) {
    return t(I18nKey.SETTINGS$MCP_ERROR_NAME_INVALID);
  }

  return null;
};

const validateNameUniqueness = (name: string): string | null => {
  if (!existingServers) {
    return null;
  }

  const shouldCheckUniqueness =
    mode === "add" || (mode === "edit" && server?.name !== name);

  if (!shouldCheckUniqueness) {
    return null;
  }

  const existingStdioNames = existingServers
    .filter((s) => s.type === "stdio")
    .map((s) => s.name)
    .filter(Boolean);

  if (existingStdioNames.includes(name)) {
    return t(I18nKey.SETTINGS$MCP_ERROR_NAME_DUPLICATE);
  }

  return null;
};

const validateCommand = (command: string): string | null => {
  if (!command) {
    return t(I18nKey.SETTINGS$MCP_ERROR_COMMAND_REQUIRED);
  }

  if (command.includes(" ")) {
    return t(I18nKey.SETTINGS$MCP_ERROR_COMMAND_NO_SPACES);
  }

  return null;
};

const validateStdioServer = (formData: FormData): string | null => {
  const name = formData.get("name")?.toString().trim();
  const command = formData.get("command")?.toString().trim();

  // Validate name
  const nameError = validateName(name || "");
  if (nameError) {
    return nameError;
  }

  // Validate name uniqueness
  const uniquenessError = validateNameUniqueness(name || "");
  if (uniquenessError) {
    return uniquenessError;
  }

  // Validate command
  const commandError = validateCommand(command || "");
  if (commandError) {
    return commandError;
  }

  return null;
};

const validateForm = (formData: FormData): string | null => {
  if (serverType === "sse" || serverType === "shttp") {
    const url = formData.get("url")?.toString().trim();
    return validateUrl(url || "");
  }

  if (serverType === "stdio") {
    return validateStdioServer(formData);
  }

  return null;
};
  ```

xingyaoww replied:
@openhands please check if this is done in this pr or not

hieptl commented on frontend/src/components/features/settings/mcp-settings/mcp-server-form.tsx:
```diff
@@ -0,0 +1,321 @@
+      }
+    }
+    return env;
+  };
```
**suggestion (non-blocking):** I think this function can be refactored as follows to avoid nested `if` statements.

```js
const parseEnvironmentVariables = (
  envString: string,
): Record<string, string> => {
  const env: Record<string, string> = {};
  if (!envString.trim()) {
    return env;
  }

  for (const line of envString.split("\n")) {
    const trimmedLine = line.trim();
    if (!trimmedLine) {
      continue;
    }

    const equalIndex = trimmedLine.indexOf("=");
    if (equalIndex === -1) {
      continue;
    }

    const key = trimmedLine.substring(0, equalIndex).trim();
    if (!key) {
      continue;
    }

    const value = trimmedLine.substring(equalIndex + 1).trim();
    env[key] = value;
  }
  return env;
};
```

xingyaoww commented:
Summary of updates addressing @hieptl's suggestions:

UI/UX:
- Server Type dropdown widths now match adjacent inputs (stdio: 350px, sse/shttp: 680px).
- Name and Server Details columns truncate with ellipsis; a single tooltip shows full content on hover.
- Header/body grid alignment updated; skeleton row grid aligned to match.

Validation:
- Environment variables field strictly enforces KEY=value per line; blank lines allowed; empty key/invalid lines rejected.
- Duplicate URL validation enforced across URL-based types (sse ↔ shttp). Please confirm whether cross-type uniqueness is desired; if you prefer same-type-only, I can change it.

Code cleanup & structure:
- Removed unused components: MCPConfigEditor and MCPJsonEditor.
- Refactored MCP server form validation into smaller functions (validateUrl, validateName, validateNameUniqueness, validateCommand, validateUrlUniqueness, validateStdioServer, parseEnvironmentVariables, formatEnvironmentVariables).

Tests & tooling:
- Added validation tests for env var parsing and cross-type duplicate URL checks: frontend/src/components/features/settings/mcp-settings/__tests__/mcp-server-form.validation.test.tsx.
- Typecheck, lint, and frontend build pass. i18n coverage is complete.

Commits:
- 679a74ff1, 2eacc22c8, 8e67d205d, 11c506543, ba6ba5282, 55d7f4213.

Follow-up:
- Cross-type duplicate policy currently enforced (sse ↔ shttp). Please confirm if this matches intent.

xingyaoww commented:
@OpenHands please fix the failing actions on PR #10145 at branch `feature/improve-mcp-config-ui`
